### DISCUSSION TITLE
nginx's prefix is /opt/homebrew/Cellar/nginx/1.25.3 but why does the file search always have html subpath ?

### DISCUSSION THREAD

**Discussion by qiulang** [Homebrew/discussions#4880] [Everyday usage] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.1.18
ORIGIN: https://github.com/Homebrew/brew
HEAD: bf8991f0d777fe8e33105f0cd7bf1e5993eed60a
Last commit: 4 days ago
Core tap JSON: 31 Oct 15:11 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit arm_blizzard_avalanche
Clang: 15.0.0 build 1500
Git: 2.39.3 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 8.1.2 => /usr/bin/curl
macOS: 14.0-arm64
CLT: 15.0.0.0.1.1694021235
Xcode: N/A
Rosetta 2: false
```

### Output of `brew doctor`

```shell
...
  /usr/local/include/node/v8config.h
  /usr/local/include/node/zconf.h
  /usr/local/include/node/zlib.h

Warning: /usr/bin occurs before /opt/homebrew/bin in your PATH.
This means that system-provided programs will be used instead of those
provided by Homebrew. Consider setting your PATH so that
/opt/homebrew/bin occurs before /usr/bin. Here is a one-liner:
  echo 'export PATH="/opt/homebrew/bin:$PATH"' >> ~/.zshrc

The following tools exist at both paths:
  openssl

Warning: Homebrew's "bin" was not found in your PATH.
Consider setting your PATH for example like so:
  echo 'export PATH="/opt/homebrew/bin:$PATH"' >> ~/.zshrc
```

### Description of issue

I checked nginx's prefix with `nginx -V` and see it is `/opt/homebrew/Cellar/nginx/1.25.3` but why does the location search always start from `/opt/homebrew/Cellar/nginx/1.25.3/html`. And I see inside this folder there is a softlink named `html` to make it work. But where does `html` part come from ?

I actually didn't notice this until I used `try_files` statement. Without the leading `/` I see error.log has the error like "[error] 40022#0: *75 open() "/opt/homebrew/Cellar/nginx/1.25.3/htmlza/static/css/chunk-libs.887691f5.css" failed " 

```
        location / {
            root   dist;
            try_files $uri $uri/ /za$uri;
        }
```

The output of `nginx -V`

```
built with OpenSSL 3.1.3 19 Sep 2023 (running with OpenSSL 3.1.4 24 Oct 2023)
TLS SNI support enabled
configure arguments: --prefix=/opt/homebrew/Cellar/nginx/1.25.3 --sbin-path=/opt/homebrew/Cellar/nginx/1.25.3/bin/nginx --with-cc-opt='-I/opt/homebrew/opt/pcre2/include -I/opt/home
...
```

**Comment by ZhongRuoyu** [comment]
The `html` part comes from [nginx's default config](https://github.com/nginx/nginx/blob/a13ed7f5ed5bebdc0b9217ffafb75ab69f835a84/conf/nginx.conf#L43-L46), which you can find at `"$(brew --prefix)"/etc/nginx/nginx.conf`:

```nginx
        location / {
            root   html;
            index  index.html index.htm;
        }
```

And `html` (`"$(brew --prefix nginx)"/html`) is a symlink to `"$(brew --prefix)"/var/www`.

If you want to change this to something else, your options include:
- make `"$(brew --prefix)"/var/www` a symlink to somewhere else;
- edit `nginx.conf` to change the `root`.

**Reply by qiulang** [reply]
Thanks for the reply. But I don't understand you last part though: "**edit nginx.conf to change the root.**" I thought I had changed it as in,

        location / {
            root   dist;
            try_files $uri $uri/ /za$uri;
        }

And actually, that was my question! I know the default value is `html` but I have changed it to `dist` so why  `try_files` still seach from `/opt/homebrew/Cellar/nginx/1.25.3/html` ?

**Reply by ZhongRuoyu** [reply]
After changing the configs, did you reload nginx with `nginx -s reload`?

**Reply by qiulang** [reply]
Of course. I reload many times in order to make `try_files` work.

**Reply by ZhongRuoyu** [reply]
I tried with your configs and added a `error_log /tmp/nginx.log debug;` directive to the `location /` configs. Then these were found in the debug logs:

<details>
<summary> <code>"$(brew --prefix)"/var/log/nginx/error.log</code> </summary>

```
2023/11/02 11:22:10 [error] 2020#0: *26 open() "/home/linuxbrew/.linuxbrew/Cellar/nginx/1.25.3/htmlza/test" failed (2: No such file or directory), client: 127.0.0.1, server: localhost, request: "GET /test HTTP/1.1", host: "localhost:8080"
```
</details>


<details>
<summary> <code>/tmp/nginx.log</code> </summary>

```
2023/11/02 11:22:10 [debug] 2020#0: *26 http cl:-1 max:1048576
2023/11/02 11:22:10 [debug] 2020#0: *26 rewrite phase: 3
2023/11/02 11:22:10 [debug] 2020#0: *26 post rewrite phase: 4
2023/11/02 11:22:10 [debug] 2020#0: *26 generic phase: 5
2023/11/02 11:22:10 [debug] 2020#0: *26 generic phase: 6
2023/11/02 11:22:10 [debug] 2020#0: *26 generic phase: 7
2023/11/02 11:22:10 [debug] 2020#0: *26 generic phase: 8
2023/11/02 11:22:10 [debug] 2020#0: *26 access phase: 9
2023/11/02 11:22:10 [debug] 2020#0: *26 access phase: 10
2023/11/02 11:22:10 [debug] 2020#0: *26 access phase: 11
2023/11/02 11:22:10 [debug] 2020#0: *26 post access phase: 12
2023/11/02 11:22:10 [debug] 2020#0: *26 generic phase: 13
2023/11/02 11:22:10 [debug] 2020#0: *26 try files handler
2023/11/02 11:22:10 [debug] 2020#0: *26 http script var: "/test"
2023/11/02 11:22:10 [debug] 2020#0: *26 trying to use file: "/test" "/home/linuxbrew/.linuxbrew/Cellar/nginx/1.25.3/dist/test"
2023/11/02 11:22:10 [debug] 2020#0: *26 http script var: "/test"
2023/11/02 11:22:10 [debug] 2020#0: *26 trying to use dir: "/test" "/home/linuxbrew/.linuxbrew/Cellar/nginx/1.25.3/dist/test"
2023/11/02 11:22:10 [debug] 2020#0: *26 http script copy: "za"
2023/11/02 11:22:10 [debug] 2020#0: *26 http script var: "/test"
2023/11/02 11:22:10 [debug] 2020#0: *26 trying to use file: "za/test" "/home/linuxbrew/.linuxbrew/Cellar/nginx/1.25.3/distza/test"
2023/11/02 11:22:10 [debug] 2020#0: *26 internal redirect: "za/test?"
```
</details>

So, as seen in `/tmp/nginx.log`, the `root dist;` config was respected, but the internal redirect at the end seems to have initiated the access to the non-existent file rooted at `html`. I don't think this is a Homebrew issue and this question may be better answered by someone who knows nginx better.

**Reply by qiulang** [reply]
As I said in my question,  homebrew clearly notices this  "The internal redirect at the end seems to have initiated the access to the non-existent file rooted at html." so it adds "html" soft link to make it.

> I see inside this folder there is a softlink named html to make it work.

**Reply by ZhongRuoyu** [reply]
> homebrew clearly notices this "The internal redirect at the end seems to have initiated the access to the non-existent file rooted at html." so it adds "html" soft link to make it

AFAICT the internal redirect is not the reason why the `html` symlink exists. It was added in https://github.com/Homebrew/legacy-homebrew/commit/f5543c372454b215f2cfda648ae875179c8ceef8 "so it is easy for people to change this on a server-by-server basis and have multiple docroots". Note that this internal redirect only happens with your `try_files` config; with the default one, *no* access to the `html` directory is made when you change `root` to somewhere else.

**Reply by qiulang** [reply]
But for my setting `root   dist;` the actual path for `dist` is inside `html`, so it dos NOT only happen for `try_files`, it happens for all.

```
(base) langqiu@langs-MacBook-Air www % ls -al /opt/homebrew/Cellar/nginx/1.25.3
total 696
drwxr-xr-x@ 16 langqiu  admin     512 Nov  1 19:59 .
drwxr-xr-x@  4 langqiu  admin     128 Nov  1 12:12 ..
-rw-r--r--@  1 langqiu  admin    6148 Nov  1 12:12 .DS_Store
drwxr-xr-x@  4 langqiu  admin     128 Oct 24 21:46 .bottle
drwxr-xr-x@  3 langqiu  admin      96 Oct 24 21:46 .brew
-rw-r--r--@  1 langqiu  admin  325032 Oct 24 21:46 CHANGES
-rw-r--r--@  1 langqiu  admin    1273 Oct 31 23:12 INSTALL_RECEIPT.json
-rw-r--r--@  1 langqiu  admin    1397 Oct 24 21:46 LICENSE
-rw-r--r--@  1 langqiu  admin      49 Oct 24 21:46 README
drwxr-xr-x@  3 langqiu  admin      96 Oct 31 23:12 bin
lrwxr-xr-x@  1 langqiu  admin      21 Nov  1 12:30 dist -> ../../../var/www/dist  
-rw-r--r--@  1 langqiu  admin     685 Oct 31 23:12 homebrew.mxcl.nginx.plist
-rw-r--r--@  1 langqiu  admin     202 Oct 31 23:12 homebrew.nginx.service
lrwxr-xr-x@  1 langqiu  admin      16 Oct 31 23:12 html -> ../../../var/www
```

See the softlink `dist -> ../../../var/www/dist ` I made (inside `html`) for `dist` to make it work here.  Without that softlink it won't work.

The `dist` actual path is, 

```
(base) langqiu@langs-MacBook-Air www % pwd
/opt/homebrew/var/www
(base) langqiu@langs-MacBook-Air www % ls
50x.html	dist		index.html	za
```

**Reply by qiulang** [reply]
I tried macport to see if there is any difference and there is! Macport behaves just like you said "this internal redirect only happens with your try_files config; with the default one, no access to the html directory is made when you change root to somewhere else."!

**Reply by ZhongRuoyu** [reply, marked as answer]
> See the softlink `dist -> ../../../var/www/dist ` I made (inside `html`) for `dist` to make it work here.

Note that a recommended way is to modify the `"$(brew --prefix)"/var/www` symlink instead (and that should work). The cellar is not preserved after an upgrade, so you may lose the `dist` symlink under the cellar directory.

> I tried macport to see if there is any difference and there is!

I don't see how the difference is; without `try_files`, Homebrew's nginx also does not access the `html` directory if you only change the root:

<details><summary>Details</summary>

```console
$ brew install nginx
[output truncated]
$ sed -Ei 's|root\s*html;|root /does/not/exist;|g' "$(brew --prefix)"/etc/nginx/nginx.conf
$ grep -C 2 root "$(brew --prefix)"/etc/nginx/nginx.conf

        location / {
            root /does/not/exist;
            index  index.html index.htm;
        }
--
        error_page   500 502 503 504  /50x.html;
        location = /50x.html {
            root /does/not/exist;
        }

--
        #
        #location ~ \.php$ {
        #    root /does/not/exist;
        #    fastcgi_pass   127.0.0.1:9000;
        #    fastcgi_index  index.php;
--
        #}

        # deny access to .htaccess files, if Apache's document root
        # concurs with nginx's one
        #
--

    #    location / {
    #        root /does/not/exist;
    #        index  index.html index.htm;
    #    }
--

    #    location / {
    #        root /does/not/exist;
    #        index  index.html index.htm;
    #    }
$ nginx
$ curl localhost:8080/test
<html>
<head><title>404 Not Found</title></head>
<body>
<center><h1>404 Not Found</h1></center>
<hr><center>nginx/1.25.3</center>
</body>
</html>
$ cat "$(brew --prefix)"/var/log/nginx/access.log
127.0.0.1 - - [06/Nov/2023:14:26:43 +0000] "GET /test HTTP/1.1" 404 153 "-" "curl/7.81.0"
$ cat "$(brew --prefix)"/var/log/nginx/error.log
2023/11/06 14:26:43 [error] 2145#0: *1 open() "/does/not/exist/test" failed (2: No such file or directory), client: 127.0.0.1, server: localhost, request: "GET /test HTTP/1.1", host: "localhost:8080"
```
</details>

**Reply by qiulang** [reply]
You are right. Both macport and homebrew versions behave the same, "this internal redirect only happens with your `try_files` config". If I only change root, there is no extra "html" in the file search path. I considered this problem solved as to why adding "html" for `try_files` is clearly not a homebrew thing.

Thanks for spending time and energy in investigating this problem.
